src/compilation.ts

```diff
@@ -0,0 +1,655 @@
+} from './utilities/graphql';
+
+import { join, block, wrap, indent } from './utilities/printing';
+
+import {
+  createHash,
+} from 'crypto';
+
+export interface CompilerOptions {
+  addTypename?: boolean;
+  mergeInFieldsFromFragmentSpreads?: boolean;
+  passthroughCustomScalars?: boolean;
+  customScalarsPrefix?: string;
+  namespace?: string;
+  generateOperationIds?: boolean;
+}
+
+export interface CompilationContext {
+  schema: GraphQLSchema;
+  operations: { [operationName: string]: CompiledOperation };
+  fragments: { [fragmentName: string]: CompiledFragment };
+  typesUsed: GraphQLType[];
+  options: CompilerOptions;
+}
+
+export interface CompiledOperation {
+  filePath?: string;
+  operationName: string;
+  operationId?: string;
+  operationType: string;
+  rootType: GraphQLObjectType;
+  variables: {
+    name: string;
+    type: GraphQLType;
+  }[];
+  source: string;
+  sourceWithFragments?: string;
+  fields: Field[];
+  fragmentSpreads?: string[];
+  inlineFragments?: CompiledInlineFragment[];
+  fragmentsReferenced: string[];
+}
+
+export interface CompiledFragment {
+  filePath?: string;
+  fragmentName: string;
+  source: string;
+  typeCondition: GraphQLCompositeType;
+  possibleTypes: GraphQLObjectType[];
+  fields: Field[];
+  fragmentSpreads: string[];
+  inlineFragments: any[];
+  fragmentsReferenced: string[];
+}
+
+export interface CompiledInlineFragment {
+  typeCondition: GraphQLObjectType;
+  possibleTypes: GraphQLObjectType[];
+  fields: Field[];
+  fragmentSpreads: string[];
+}
+
+export interface Field {
+  responseName: string;
+  fieldName: string;
+  args?: Argument[];
+  type: GraphQLType;
+  description?: string;
+  directives?: DirectiveNode[];
+  selectionSet?: SelectionSetNode;
+  isConditional?: boolean;
+  isDeprecated?: boolean;
+  deprecationReason?: string;
+  fields?: Field[];
+  fragmentSpreads?: string[];
+  inlineFragments?: CompiledInlineFragment[];
+}
+
+export interface Argument {
+  name: string;
+  value: any;
+}
+
+type GroupedFieldSet = Map<string, FieldSet>;
+type FieldSet = [GraphQLCompositeType, Field][];
+
+type VisitedFragmentSet = Set<string>;
+type GroupedVisitedFragmentSet = Map<GraphQLCompositeType, VisitedFragmentSet>;
+type FragmentsReferencedSet = Set<string>;
+
+// Parts of this code are adapted from graphql-js
+
+export function compileToIR(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  options: CompilerOptions = { mergeInFieldsFromFragmentSpreads: true }
+): CompilationContext {
+  if (options.addTypename) {
+    document = withTypenameFieldAddedWhereNeeded(schema, document);
+  }
+
+  const compiler = new Compiler(schema, document, options);
+
+  const operations: {
+    [operationName: string]: CompiledOperation;
+  } = Object.create(null);
+
+  compiler.compiledOperations.forEach(operation => {
+    operations[operation.operationName] = operation;
+  });
+
+  const fragments: { [fragmentName: string]: CompiledFragment } = Object.create(null);
```

Comment:
The main reason was to stay compatible with the current compiler output, because that is what the targets depend on. It also makes it easier to stringify as JSON (for Android), but there are other ways to get around that.